normalize glob endings

ABaldwinHunter opened the pull request:
@codeclimate/review PR:

1) Normalizes directory endings for paths => `/**` => `**/*`
2) and `**.rb` => `**/*.rb`
3) Removes code for `rate?` and `exclude?` methods which will no longer be responsibility of `cc/yaml`. codeclimate-yaml will only be responsible for culling data from config file.

GordonDiggs commented on lib/cc/yaml/nodes/glob.rb:
```diff
@@ -2,9 +2,6 @@ module CC
   module Yaml
     module Nodes
       class Glob < Scalar
```
Do we still need this and `GlobList`?

ABaldwinHunter replied:
Ah no I think not. Nice

pbrisbin replied:
I think you need `GlobList` to map `ratings.paths` to a `Sequence` specifically of `Glob`s... it holds the `type` declaration.

ABaldwinHunter replied:
I think it does this already? https://github.com/codeclimate/codeclimate-yaml/blob/abh_normalize_glob_endings/lib/cc/yaml/nodes/ratings.rb#L5

pbrisbin replied:
Maybe I'm misunderstanding this commen then. You can ignore me.

ABaldwinHunter commented:
dog food comment: it feels slightly deflating to have `162` fixed issues and `1` new, but still get a red color on cc check :( (even though in this case 'fixes' are just from disabling some rubocop checks). Is that our intended behavior?

GordonDiggs commented:
@ABaldwinHunter yep any new issues fail the PR, regardless of how many others you fix

ABaldwinHunter commented:
ah okay.

GordonDiggs commented on lib/cc/yaml/nodes/sequence.rb:
```diff
@@ -113,10 +117,21 @@ def add_value!(value)
+
+      private
+
+      def normalize_path_endings(value)
```
This seems to be the odd functionality out in this class. What do you think of extracting it to its own `PathNormalizer` class?

GordonDiggs commented on lib/cc/yaml/nodes/sequence.rb:
```diff
@@ -113,10 +117,21 @@ def add_value!(value)
+
+      private
+
+      def normalize_paths(value)
```
Reviving my comment about pulling this out. What do you think?

pbrisbin commented:
It seems weird to remove Glob and push this into Sequence. Why not keep Glob and have it be in charge of converting itself? I think overriding `#value` is an option.

ABaldwinHunter commented:
@GordonDiggs I like the `PathNormalizer` class idea. @pbrisbin I could try to implement that. It was my first attempt but seemed a bit trickier to implement. I might have just been missing something though. Is our consensus that that'd be preferable?

GordonDiggs commented:
I think overriding `value` would achieve the goal of my comment (probably even better)

ABaldwinHunter commented:
Okay sounds good. Switching to that approach.

ABaldwinHunter commented:
@codeclimate/review Ready for re-review!

pbrisbin commented on lib/cc/yaml/nodes/sequence.rb:
```diff
@@ -113,10 +113,10 @@ def add_value!(value)
+      def dup_values
+        @children = @children.map { |child| child.dup }
+        self
+      end
```
I would revert this just to keep the diff cleaner. Our rule is not to change code *just* to change style.

ABaldwinHunter replied:
Okay. Reverted!

pbrisbin commented on spec/cc/yaml/nodes/glob_spec.rb:
```diff
@@ -0,0 +1,108 @@
+
+  describe "value" do
+    it "normalizes path" do
+
```
What's this?

pbrisbin commented on lib/cc/yaml/nodes/glob.rb:
```diff
@@ -2,8 +2,11 @@ module CC
-        def match?(path)
-          File.fnmatch(to_s, path)
+        DOUBLE_STAR_END = Regexp.new('\/\*\*$').freeze
+        DOUBLE_STAR_BEGIN = Regexp.new('\*\*\.').freeze
```
I don't see a ton of value in these constants. Personally, I'd inline them.

:lipstick: our style guide says `%r{ }` for regexes with slashes in it.

pbrisbin commented on lib/cc/yaml/nodes/glob.rb:
```diff
@@ -2,8 +2,11 @@ module CC
+        DOUBLE_STAR_BEGIN = Regexp.new('\*\*\.').freeze
+
+        def value
+          @value.sub(DOUBLE_STAR_END, "/**/*").sub(DOUBLE_STAR_BEGIN, "**/*.")
```
Did you spend and time trying to do this with a single `sub`? Now that we have all those (great) test cases, I might much around with it a little bit :)

pbrisbin commented:
This LGTM. Remember that we can merge, but not ship the updated gem except in a CLI/Builder that correctly handles the `**/*` globs.

ABaldwinHunter commented:
@pbrisbin thanks for comments! I updated regex - removed constants.

I ran into a block trying to make one substitution, mostly because:
1) the item being matched is different (`**.` and `**$`). Even when I could make a regex that wouldmatch either situation, I think the part we'd want to swap out would ONLY be the `**`.

So let me know if you have a different idea! :)

pbrisbin commented:
Frustrating yourself with regex when you've got something that works is definitely not worth the time, so merging this as-is would be totally fine. But...

```rb
@value.sub(/\*\*(\.[^\/]*)?$/, "**/*\\1")
```

passes your specs. Basically it's the `**` pattern, an optional extension, then the end of the string, with the replacement being `**/*` plus the extension (which may be nothing if one wasn't present).

This regex does allow empty extensions (so `**.` will become `**/*.`), and it won't expand `**` pieces that aren't the last (so `x/**/y` will not become `x/**/*/y`). I think both of these behaviors are what we want.

Also, this line is probably a good use case for a commit :)

pbrisbin commented:
On second thought, I'm not sure we need to treat an extension any specially, we'd want `**foo` to expand to `**/*foo` right? if so, then this is what I'd use:

```rb
@value.sub(/\*\*([^\/]*)?$/, "**/*\\1")
```

It also passes the current specs.

ABaldwinHunter commented:
@pbrisbin awesome. Thanks for the second pass! I like the way you made the regex specifically exclude the `\/` character after `**` instead of checking for the positive matches of `**$` or `**.`. You're right; not `/` is pretty much the only case we need to avoid hitting a match.

I updated regex and added inline comment.
